src/main/java/com/minecolonies/coremod/network/messages/BuildingMoveMessage.java

```diff
@@ -0,0 +1,254 @@
+        if (!Structures.hasMD5(sn))
+        {
+            player.sendMessage(new TextComponentString("Can not build " + message.workOrderName + ": schematic missing!"));
+            return;
+        }
+        handleHut(CompatibilityUtils.getWorld(player), player, sn, message.rotation, message.pos, message.mirror, message.buildingId);
+    }
+
+    /**
+     * Handles the placement of huts.
+     *
+     * @param world         World the hut is being placed into.
+     * @param player        Who placed the hut.
+     * @param sn            The name of the structure.
+     * @param rotation      The number of times the structure should be rotated.
+     * @param buildPos      The location the hut is being placed.
+     * @param mirror        Whether or not the strcture is mirrored.
+     * @param oldBuildingId The old building id.
+     */
+    private static void handleHut(
+            @NotNull final World world, @NotNull final EntityPlayer player,
+            final StructureName sn,
+            final int rotation, @NotNull final BlockPos buildPos, final boolean mirror, final BlockPos oldBuildingId)
+    {
+        final String hut = sn.getSection();
+        final Block block = Block.getBlockFromName(Constants.MOD_ID + ":blockHut" + hut);
+        final Colony tempColony = ColonyManager.getClosestColony(world, buildPos);
+        if (tempColony != null
+                && (!tempColony.getPermissions().hasPermission(player, Action.MANAGE_HUTS)
+                && !(block instanceof BlockHutTownHall
+                && !ColonyManager.isTooCloseToColony(world, buildPos))))
+        {
+            return;
+        }
+
+        @Nullable AbstractBuilding oldBuilding = ColonyManager.getBuilding(world, oldBuildingId);
+        if (oldBuilding instanceof BuildingTownHall)
+        {
+            oldBuilding = null;
+            world.setBlockToAir(oldBuildingId);
+        }
+
+        if (block != null && EventHandler.onBlockHutPlaced(world, player, block, buildPos))
+        {
+            world.destroyBlock(buildPos, true);
+            world.setBlockState(buildPos, block.getDefaultState().withRotation(BlockUtils.getRotation(rotation)));
+            ((AbstractBlockHut) block).onBlockPlacedByBuildTool(world, buildPos, world.getBlockState(buildPos), player, null, mirror, sn.getStyle());
+            setupBuilding(world, player, sn, rotation, buildPos, mirror, oldBuilding);
+        }
+    }
+
+    /**
+     * setup the building once it has been placed.
+     *  @param world         World the hut is being placed into.
+     * @param player        Who placed the hut.
+     * @param sn            The name of the structure.
+     * @param rotation      The number of times the structure should be rotated.
+     * @param buildPos      The location the hut is being placed.
+     * @param mirror        Whether or not the strcture is mirrored.
+     * @param oldBuilding The old building id.
+     */
+    private static void setupBuilding(
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 28 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)